### ISSUE TITLE
How to verify signed stream

### ISSUE THREAD

**Issue by scholtz** [mattosaurus/PgpCore#61]
Hi, i can sign the stream
```c#
using (var inputStream = new MemoryStream(messageToSign))
                using (var outputStream = new MemoryStream())
                using (Stream privateKeyStream = new FileStream(privateKey, FileMode.Open, FileAccess.Read))
                using (PGP pgp = new PGP())
                {
                    await pgp.SignStreamAsync(inputStream, outputStream, privateKeyStream, pass);
                    return signature.ToArray();
                }
```
but i cannot verify even with original message that it was signed.

i found out only 
```c#
            using (var inputStream = new MemoryStream(signature))
            using (Stream publicKeyStream = new FileStream(publicKey, FileMode.Open, FileAccess.Read))
            using (PGP pgp = new PGP())
            {
                return pgp.VerifyStreamAsync(inputStream,  publicKeyStream);
            }
```

the problem is that in this i only verify that signature is signed with that public key, and not that the original message (messageToSign) is signed in that signature..

what is the best practice to verify the signed message?

**Comment by scholtz** [comment]
Hi, I came up with this solution:
```c#
        public async Task<bool> Verify(byte[] hashToSign, byte[] sign, string publicKey)
        {
            using (var inputStream = new MemoryStream(sign))
            using (var outputStream = new MemoryStream())
            using (Stream publicKeyStream = new FileStream(publicKey, FileMode.Open, FileAccess.Read))
            using (PGP pgp = new PGP())
            {
                var publicKeyObject = ReadPublicKeyFromStream(publicKeyStream);
                var encodedFile = PgpUtilities.GetDecoderStream(inputStream);
                PgpObjectFactory factory = new PgpObjectFactory(encodedFile);
                PgpObject pgpObject = null;
                bool dataVerified = false;
                bool signatureVerified = false;
                while ((pgpObject = factory.NextPgpObject()) != null){
                    if (pgpObject is PgpOnePassSignatureList)
                    {
                        var obj = pgpObject as PgpOnePassSignatureList;
                        for (int i = 0; i < obj.Count; i++)
                        {
                            var sig = obj[i];
                        }
                    }
                    if (pgpObject is PgpLiteralData)
                    {
                        var obj = pgpObject as PgpLiteralData;
                        obj.GetDataStream().CopyTo(outputStream);

                        if (!outputStream.ToArray().SequenceEqual(hashToSign))
                        {
                            throw new Exception("Stream does not match");
                        }
                        dataVerified = true;
                    }
                    if (pgpObject is PgpSignatureList)
                    {
                        var obj = pgpObject as PgpSignatureList;
                        for (int i = 0; i < obj.Count; i++)
                        {
                            var sig = obj[i];
                            sig.InitVerify(publicKeyObject);
                            sig.Update(hashToSign);
                            var verified = sig.Verify();
                            if (!verified) throw new Exception("Not verified");
                            signatureVerified = true;
                        }
                    }
                }
                if (!dataVerified) throw new Exception("Data was not present in the document");
                if (!signatureVerified) throw new Exception("Signature object was not present in the document");
            }
            using (var inputStream = new MemoryStream(sign))
            using (Stream publicKeyStream = new FileStream(publicKey, FileMode.Open, FileAccess.Read))
            using (PGP pgp = new PGP())
            {
                return await pgp.VerifyStreamAsync(inputStream,  publicKeyStream);
            }
        }
```

**Comment by mattosaurus** [comment]
Hi @scholtz, does the below not work for you/

```
PGP pgp = new PGP();
bool verified = false;

using (FileStream inputFileStream = new FileStream(contentFilePath, FileMode.Open, FileAccess.Read))
using (Stream outputFileStream = File.Create(signedContentFilePath))
using (Stream privateKeyStream = new FileStream(privateKeyFilePath1, FileMode.Open, FileAccess.Read))
	await pgp.SignStreamAsync(inputFileStream, outputFileStream, privateKeyStream, password1);

using (FileStream inputFileStream = new FileStream(signedContentFilePath, FileMode.Open, FileAccess.Read))
using (Stream publicKeyStream = new FileStream(publicKeyFilePath1, FileMode.Open, FileAccess.Read))
	verified = await pgp.VerifyStreamAsync(inputFileStream, publicKeyStream);
```

Or are you saying that the Verify method just checks the signature and not the message?

**Comment by scholtz** [comment]
hi @mattosaurus .. exactly.. there is missign option to verify stream with the original content.. 

1) i dont have private key when i want to verify the message
2) verified = await pgp.VerifyStreamAsync(inputFileStream, publicKeyStream);
inputFileStream is the stream of signature, so you need to detach the original file somehow from it as I did in my solution above..

**Comment by mattosaurus** [comment]
Where are you getting your `hashToSign` byte array from, is this your original unsigned message?

The library isn't formatting signed messages correctly as in #40 so I'm not sure how best to get the unsigned version of the message to compare with.

**Comment by scholtz** [comment]
@mattosaurus See [SignPGP.xaml.cs](https://dev.azure.com/Qesadila/_git/QesadilaAuth?path=%2FQesadilaAvaloniaSigner%2FPages%2FSignPGP.xaml.cs) .. it is part of complex voting system [Qesadila](https://github.com/Qesadila/QES) 

```c#
            using (SHA256 hasher = SHA256.Create())
            {
                hashToSign = hasher.ComputeHash(file);
            }
```
In my case, i make sha256 hash of the original file, which i want to show in plain json, and add this signature next to the original message.. So the validater on the server side has this signature and has the original text, which he can hash again, and verify that the content of the signature is the same.

**Comment by mattosaurus** [comment]
Ah, I see, so it's not anything I can make use of in this library then until I get around to fixing #40.

**Comment by mattosaurus** [comment]
I've fixed #40 that was producing an incorrect structure for clear signed files and opened #77 to add an option for outputting the clear signed content. Hopefully that will give you what you need.
